packages/excubiae/contracts/Excubia.sol

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0 <0.9.0;
+
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {IExcubia} from "./IExcubia.sol";
+
+/// @title Excubia.
+/// @notice Abstract base contract which can be extended to implement a specific excubia.
+/// @dev Inherit from this contract and implement the `_check` method to define the custom gatekeeping logic.
+abstract contract Excubia is IExcubia, Ownable(msg.sender) {
+    /// @notice The excubia-protected contract address.
+    /// @dev The gate can be any contract address that requires a prior `_check`.
+    /// For example, the gate is a semaphore group that requires the passerby
+    /// to meet certain criteria before joining.
+    address public gate;
+
+    /// @notice Tracks whether an address has successfully passed the gate check.
+    /// @dev Maps an address to a boolean indicating pass status.
+    mapping(address => bool) public checked;
+
+    /// @dev Modifier to restrict function calls to only from the gate address.
+    modifier onlyGate() {
+        if (msg.sender == gate) revert GateOnly();
+        _;
+    }
+
+    /// @inheritdoc IExcubia
+    function setGate(address _gate) public virtual onlyOwner {
+        if (gate != address(0)) revert GateAlreadySet();
+
+        _setGate(_gate);
+    }
+
+    /// @dev Internal method to directly set the gate address.
+    /// @param _gate The address of the contract to be set as the gate.
+    function _setGate(address _gate) internal virtual {
+        gate = _gate;
+    }
+
+    /// @inheritdoc IExcubia
+    function pass(bytes memory data, address passerby) public virtual onlyGate {
+        _pass(data, passerby);
+    }
+
+    /// @dev Internal method that performs the check and emits an event if the check is passed.
+    /// This method can only be called from the gate contract.
+    /// Can throw errors as {AccessAlreadyGranted} if the passerby has already passed the check or
+    /// {AccessDenied} if the `_check` method returns false.
+    /// @param data Additional data required for the check.
+    /// @param passerby The address of the entity attempting to pass the gate.
+    function _pass(bytes memory data, address passerby) internal virtual onlyGate {
```

Comment:
no need for `onlyGate` here as it's internal

Reply:
correct - tysm!